zebedee-cms/src/main/java/com/github/onsdigital/zebedee/keyring/KeyringImpl.java

```diff
@@ -102,30 +101,84 @@ public SecretKey get(User user, Collection collection) throws KeyringException {
 
     @Override
     public void remove(User user, Collection collection) throws KeyringException {
-        // TODO
+        validateUser(user);
+        validateUserEmail(user);
+        validateCollection(collection);
+
+        boolean hasAccess = false;
+        try {
+            hasAccess = permissionsService.canEdit(user.getEmail());
+        } catch (IOException ex) {
+            throw new KeyringException(ex);
+        }
+
+        if (!hasAccess) {
+            return;
+        }
+
+        cache.remove(collection.getDescription().getId());
     }
 
     @Override
     public void add(User user, Collection collection, SecretKey key) throws KeyringException {
-        // TODO
+        validateUser(user);
+        validateUserEmail(user);
+        validateCollection(collection);
+        validateKey(key);
+
+        boolean hasAccess = false;
+        try {
+            hasAccess = permissionsService.canEdit(user.getEmail());
+        } catch (IOException ex) {
+            throw new KeyringException(ex);
+        }
+
+        if (!hasAccess) {
+            return;
+        }
+
+        cache.add(collection.getDescription().getId(), key);
     }
 
-    /**
-     * Lists the collection IDs in the keyring.
-     *
-     * @param user
-     * @return An unmodifiable set of the key identifiers in the keyring.
-     */
+
     @Override
     public Set<String> list(User user) throws KeyringException {
-        return null;
-    }
+        validateUser(user);
+        validateUserEmail(user);
+        boolean hasAccess = false;
+        try {
+            hasAccess = permissionsService.canEdit(user.getEmail());
+        } catch (IOException ex) {
+            throw new KeyringException(ex);
+        }
 
-    private void validGetParams(User user, Collection collection) throws KeyringException {
+        if (!hasAccess) {
+            return null;
+        }
+
+        return cache.list();
+    }
+    
+    private void validateUser(User user) throws KeyringException {
         if (user == null) {
             throw new KeyringException(USER_NULL_ERR);
         }
+    }
+
+    private void validateUserEmail(User user) throws KeyringException {
+        if (user.getEmail()== null | user.getEmail()== "") {
+            throw new KeyringException(USER_EMAIL_ERR);
+        }
+    }
```

Comment:
I think this should be part of the `validateUser()` method as the email is a property of the user so it makes sense that validating the user should check this field.

Should be able to cut n paste this logic into the validateUser method.